src/timepicker/TimePicker.ts

```diff
@@ -0,0 +1,481 @@
+
+	render(): DNode {
+		const { disabled, formId, invalid, label, readOnly, required } = this.properties;
+		const options = this.renderOptions();
+		const menuId = options ? (<ResultMenuProperties> options.properties).id : '';
+
+		let controls: DNode = v('span', {
+			classes: this.classes(css.controls)
+		}, [
+			this.renderCustomInput(),
+			this.renderButton(menuId as string)
+		]);
+
+		if (label) {
+			controls = w(Label, {
+				bind: this,
+				classes: this.classes(
+					disabled ? css.disabled : null,
+					invalid ? css.invalid : null,
+					readOnly ? css.readOnly : null,
+					required ? css.required : null
+				),
+				formId,
+				label
+			}, [ controls ]);
+		}
+
+		return v('span', {
+			'aria-expanded': this._open ? 'true' : 'false',
+			'aria-haspopup': 'true',
+			'aria-readonly': readOnly ? 'true' : 'false',
+			'aria-required': required ? 'true' : 'false',
+			classes: this.classes(css.root),
+			key: 'root'
+		}, [
+			controls,
+			options
+		]);
+	}
+
+	protected onElementCreated(element: HTMLElement, key: string) {
+		if (key === 'root') {
+			this._inputElement = <HTMLInputElement> element.querySelector('input');
+			this._restoreFocus();
+		}
+	}
+
+	protected onElementUpdated(element: HTMLElement, key: string) {
+		if (key === 'root') {
+			this._restoreFocus();
+			const selectedResult = element.querySelector('[data-selected="true"]');
+			selectedResult && this._scrollIntoView(<HTMLElement> selectedResult);
+		}
+	}
+
+	@onPropertiesChanged()
+	protected onPropertiesChanged(evt: PropertiesChangeEvent<this, TimePickerProperties>) {
+		const {
+			customOptionItem = ResultItem,
+			customOptionMenu = ResultMenu
+		} = this.properties;
+
+		if (
+			!this.registry ||
+			includes(evt.changedPropertyKeys, 'customOptionItem') ||
+			includes(evt.changedPropertyKeys, 'customOptionMenu')) {
+			this.registry = this._createRegistry(customOptionItem, customOptionMenu);
+		}
+	}
+
+	protected renderButton(menuId: string) {
+		const { buttonText, disabled, readOnly } = this.properties;
+
+		return v('button', {
+			'aria-controls': menuId,
+			bind: this,
+			classes: this.classes(css.button),
+			disabled,
+			onclick: this._onButtonClick,
+			readOnly
+		}, buttonText ? [ buttonText ] : undefined);
+	}
+
+	protected renderCustomInput() {
+		const {
+			disabled,
+			inputProperties = {},
+			invalid,
+			readOnly,
+			required,
+			value
+		} = this.properties;
+
+		return w(TextInput, {
+			...inputProperties,
+			bind: this,
+			disabled,
+			invalid,
+			onBlur: this._onInputBlur,
+			onFocus: this._onInputFocus,
+			onInput: this._onInput,
+			onKeyDown: this._onInputKeyDown,
+			overrideClasses: css,
+			readOnly,
+			required,
+			value
+		});
+	}
+
+	protected renderOptions() {
+		if (!this._open) {
+			return null;
+		}
+
+		const { isOptionDisabled, options = [] } = this.properties;
+		return w('result-menu', <ResultMenuProperties> {
+			bind: this,
+			getResultLabel: this._getOptionLabel,
+			id: uuid(),
```

Comment:
Not sure how much we care about this, but calling `uuid()` here will change the `id` every single render call, and require dom changes in both this and the button.